synapse/federation/federation_client.py

```diff
@@ -345,6 +346,83 @@ def claim_client_keys(self, destination, content, timeout):
 
         return state_event_ids, auth_event_ids
 
+    async def _check_sigs_and_hash_and_fetch(
+        self,
+        origin: str,
+        pdus: List[EventBase],
+        room_version: str,
+        outlier: bool = False,
+        include_none: bool = False,
+    ) -> List[EventBase]:
+        """Takes a list of PDUs and checks the signatures and hashs of each
+        one. If a PDU fails its signature check then we check if we have it in
+        the database and if not then request if from the originating server of
+        that PDU.
+
+        If a PDU fails its content hash check then it is redacted.
+
+        The given list of PDUs are not modified, instead the function returns
+        a new list.
+
+        Args:
+            origin
+            pdu
+            room_version
+            outlier: Whether the events are outliers or not
+            include_none: Whether to include None in the returned list
+                for events that have failed their checks
+
+        Returns:
+            Deferred : A list of PDUs that have valid signatures and hashes.
+        """
+        deferreds = self._check_sigs_and_hashes(room_version, pdus)
+
+        @defer.inlineCallbacks
+        def handle_check_result(pdu: EventBase, deferred: Deferred):
+            try:
+                res = yield make_deferred_yieldable(deferred)
+            except SynapseError:
+                res = None
+
+            if not res:
+                # Check local db.
+                res = yield self.store.get_event(
+                    pdu.event_id, allow_rejected=True, allow_none=True
+                )
+
+            if not res and pdu.origin != origin:
+                try:
+                    res = yield defer.ensureDeferred(
+                        self.get_pdu(
+                            destinations=[pdu.origin],
+                            event_id=pdu.event_id,
+                            room_version=room_version,  # type: ignore
```

Comment:
Without ignoring this for mypy it complains:

```
synapse/federation/federation_client.py:399: error: Argument "room_version" to "get_pdu" of "FederationClient" has incompatible type "str"; expected "RoomVersion"  [arg-type]
```

I attempted to fix this in e5e5b4c7846650cb6c19f4e563e3c4312973b53c, but that seems to break SyTest. I'm not 100% confident whether mypy is in error or the code is, although I did trace through it and it seems like the code is wrong. 🤷‍♂

Reply:
(Or maybe the typing is wrong on one of these functions.)

Reply:
uh yeah, mypy has a point: this doesn't look right.

It's a bit of an edge-case so I can believe it's been broken for a few weeks without us noticing. Looks like it got broken in f84700fba8cd9345d7a1a025462c7f8650f27386, for the same reason that 0cb0c7bcd529b0dda92aff0dae6277dfcb6f6e29: I didn't find this use of `get_pdu` when searching.

To fix it, it looks like it should be easy enough to pass a `RoomVersion` object into `_check_sigs_and_hash_and_fetch` instead of a str? Don't know if you want to do that in this PR or another.

Reply:
I'd like to do it in a follow-up if that's OK. I *think* this is what my commit above does and it ends up being a bit more invasive than expected.

Reply:
See #7089.